bdsync for windows server

Hi Everyone, this is not exactly an issue but a feature/port request.

I have tested bdsync on Linux and it runs great and I could actually sync a server's block device into an image file which I then imported into a VM in a different location.

Would it be possible to do this for Windows servers while the servers are still powered on?
If it is possible what would it require?
My plan is to build a disaster recovery / business continuity solution around bdsync and would like to know if the same software can be ported/written to windows

Thanks

Unfortunatly it doesn't compile under cygwin. Would love to use bdsync for syncing VM-Images on my Windows Desktop.

```
In file included from bdsync.c:56:
bdsync.c: In function ‘handle_err’:
bdsync.c:435:26: warning: array subscript has type ‘char’ [-Wchar-subscripts]
  435 |             if (!isprint(*rp)) {
      |                          ^~~
bdsync.c: In function ‘opendev’:
bdsync.c:1724:29: error: ‘O_LARGEFILE’ undeclared (first use in this function)
 1724 |     fd = open (dev, flags | O_LARGEFILE);
      |                             ^~~~~~~~~~~
```

--- a/.gitignore
+++ b/.gitignore
@@ -1,4 +1,8 @@
 bdsync
+bdsync.exe
 bdsync*.tgz
 bdsync.1
 bdsync.txt.2
+*~
+*.swp
+

--- a/COMPILING
+++ b/COMPILING
@@ -10,7 +10,7 @@ sudo apt install build-essential checkinstall wget
 sudo apt install libssl-dev pandoc 
 # download and unzip bdsync
 wget https://github.com/rolffokkens/bdsync/archive/master.zip
-unzip bdsync.zip
+unzip master.zip
 # compile bdsync
 cd bdsync-master/
 make
@@ -23,8 +23,48 @@ CentOS/RedHat/Fedora
 sudo yum install openssl-devel gcc pandoc wget
 # download and unzip bdsync
 wget https://github.com/rolffokkens/bdsync/archive/master.zip
-unzip bdsync.zip
+unzip master.zip
 # compile bdsync
 cd bdsync-master/
 make
 # done -- you'll find bdsync in current path
+
+Windows (using cygwin)
+======================
+
+# install cygwin for Windows
+https://cygwin.com/install.html
+# Select packages gcc-g++, make, openssl-devel, wget, unzip on the 'Select Packages' screen
+# Install Pandoc
+https://pandoc.org/installing.html
+# Add pandoc to the path
+echo 'export PATH="$PATH:/cygdrive/c/program files/pandoc"' > /etc/profile.d/pandoc.sh
+source /etc/profile.d/pandoc.sh
+# download and unzip bdsync
+wget https://github.com/rolffokkens/bdsync/archive/master.zip
+unzip master.zip
+# compile bdsync
+cd bdsync-master/
+make
+# done -- you'll find bdsync in current path
+
+Setting up for Windows (using cygwin)
+=====================================
+
+# Setting up SSH on Windows machines:
+# Re-run setup-x86_64.exe of cygwin
+# Install package openssh
+# Run Cygwin terminal as administrator, run command
+ssh-host-config
+# To allow sshd in the firewall, execute the following in PowerShell as administrator:
+New-NetFirewallRule -Name sshd -DisplayName 'CYGWIN sshd' -Enabled True -Direction Inbound -Protocol TCP -Action Allow -LocalPort 22 -Program "C:\apps\cygwin64\usr\sbin\sshd.exe"
+# Start the SSHD service in Cygwin terminal
+net start cygsshd
+# If key authenication is required, generate a new private / public pair using ssh-keygen
+# Public key should be placed in ~/.ssh/authorized_keys on the server
+# Private key should be defined in ~/.ssh/config on the client:
+Host <remotehost>
+  IdentityFile c:/path/<yourprivatekey>
+# The permissions on the private key should be restricted (unless StrictModes in sshd is off):
+# Properties / Security / Advanced / Disable Inheritance. Make sure only two users remain listed: SYSTEM and your Windows username
+# To debug connections, enable logging for SSH: install syslog-ng and update SysLogFacility / LogLevel in /etc/sshd_config

--- a/Makefile
+++ b/Makefile
@@ -10,7 +10,11 @@ CFLAGS=-O3
 all: bdsync bdsync.1
 
 bdsync.txt.2: bdsync.txt
-	sed 's/^/"/;s/$$/\\n"/' bdsync.txt > bdsync.txt.2
+ifndef OS
+	sed 's/\(.*\)/"\1\\n"/g' bdsync.txt > bdsync.txt.2
+else
+	sed 's/\(.*\)\r/"\1\\r\\n"/g' bdsync.txt > bdsync.txt.2
+endif
 
 bdsync: bdsync.c checkzero.c bdsync.txt.2
 	$(CC) -Wall $(CFLAGS) $(CRYPTO_DEF) -o bdsync bdsync.c checkzero.c $(CRYPTO_LDFLAGS)
@@ -25,4 +29,4 @@ test: bdsync
 	./tests.sh
 
 clean:
-	rm -f bdsync
+	rm -f bdsync bdsync.txt.2 bdsync.1

--- a/bdsync.c
+++ b/bdsync.c
@@ -422,7 +422,7 @@ int handle_err (int fd)
                 /* if (ep == ebuf + EBUFSIZ) ep = ebuf; */
                 ep = ebuf;
             }
-            if (!isprint(*rp)) {
+            if (!isprint((int) *rp)) {
                 rp++;
                 sz--;
                 continue;
@@ -1711,7 +1711,11 @@ struct dev *opendev (char *dev, int flags, int flushcache)
     int        fd;
     struct dev *devp;
 
+#ifdef O_LARGEFILE
     fd = open (dev, flags | O_LARGEFILE);
+#else
+    fd = open (dev, flags);
+#endif
     if (fd == -1) {
         verbose (0, "opendev [%s]: %s\n", dev, strerror (errno));
         exit (exitcode_io_error);

